core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,537 @@
+         //nothing to release
+         return;
+      }
+
+      final boolean released = wantToRelease.setReleased();
+      if (trace) {
+         log.tracef("Release lock for %s? %s", wantToRelease, released);
+      }
+
+      LockPlaceHolder currentLocked = current;
+      if (currentLocked == wantToRelease) {
+         if (!casRelease(currentLocked)) {
+            if (trace) {
+               log.tracef("Releasing lock for %s. It is the current lock owner but another thread changed it.", lockOwner);
+            }
+            //another thread already released!
+         } else {
+            tryAcquire();
+         }
+      }
+   }
+
+   /**
+    * @return {@code true} if this lock has no lock owner pending.
+    */
+   public boolean isEmpty() {
+      return isFree() && lockOwners.isEmpty();
+   }
+
+   /**
+    * @return {@code true} if the lock is not acquired and no lock owners are queued.
+    */
+   public boolean isFree() {
+      return current == null && pendingRequest.isEmpty();
+   }
+
+   /**
+    * @return the current lock owner or {@code null} if it is not acquired.
+    */
+   public Object getLockOwner() {
+      LockPlaceHolder lockPlaceHolder = current;
+      return lockPlaceHolder == null ? null : lockPlaceHolder.owner;
+   }
+
+   /**
+    * It checks if the lock is acquired.
+    * <p/>
+    * A {@code false} return value does not mean the lock is free since it may have queued lock owners.
+    *
+    * @return {@code true} if the lock is acquired.
+    */
+   public boolean isLocked() {
+      return current != null;
+   }
+
+   /**
+    * It forces a deadlock checking.
+    */
+   public void deadlockCheck() {
+      LockPlaceHolder holder = current;
+      if (holder != null) {
+         for (LockPlaceHolder pending : pendingRequest) {
+            pending.checkDeadlock(holder.owner);
+         }
+      }
+   }
+
+   /**
+    * It tests if the lock has the lock owner.
+    * <p/>
+    * It return {@code true} if the lock owner is the current lock owner or it in the queue.
+    *
+    * @param lockOwner the lock owner to test.
+    * @return {@code true} if it contains the lock owner.
+    */
+   public boolean containsLockOwner(Object lockOwner) {
+      return lockOwners.containsKey(lockOwner);
+   }
+
+   private void onCanceled(LockPlaceHolder canceled) {
+      if (trace) {
+         log.tracef("Release lock for %s. It was canceled.", canceled.owner);
+      }
+      LockPlaceHolder wantToRelease = lockOwners.get(canceled.owner);
+      if (wantToRelease == null) {
+         if (trace) {
+            log.tracef("%s not found!", canceled.owner);
+         }
+         //nothing to release
+         return;
+      }
+      LockPlaceHolder currentLocked = current;
+      if (currentLocked == wantToRelease) {
+         if (!casRelease(currentLocked)) {
+            if (trace) {
+               log.tracef("Releasing lock for %s. It is the current lock owner but another thread changed it.", canceled.owner);
```

Comment:
The originator won't retry until it receives a response from all the owners, so a topology change can't cause this. A timeout during prepare/commit may end up with the prepare/commit and the rollback running at the same time, but `TransactionSynchronizerInterceptor` will block the rollback until the prepare finishes. Unless you can think of another way this could happen, I would like to throw an exception here.